### PR TITLE
176 add apple health

### PR DESCRIPTION
**Pull request by Orta21**
Ticket: #51 

### Description

Apple health integration

### Release Plan

Once approved

### CODE REVIEW

**Comment by leite08** [comment]
I suggest we update the commit messages (or at least the last one) following Conventional Commits so we can automatically identify the type of changes and determine the next version using semantic versioning - [as discussed here](https://metriport.slack.com/archives/C04DMKE9DME/p1673049721707849)

image

File: connect-widget/app/.env.development [hunk 1]
```diff
@@ -1,3 +1,3 @@
 # Backend server endpoint
 PORT=3001
-REACT_APP_API_URL=http://0.0.0.0:8080
\ No newline at end of file
+REACT_APP_API_URL=http://192.168.0.139:8080
```
Comment (Goncharo) [inline connect-widget/app/.env.development:3]: Why was this changed? should be `0.0.0.0` ?

**Review by Goncharo — CHANGES_REQUESTED** [review]
Review still in progress - posting comments so far

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthStore.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    public let healthStore = HKHealthStore()
+    public let metriportClient: MetriportClient
+    private let healthKitTypes = HealthKitTypes()
+    private var metriportUserId = ""
+
+    public init() {
+        self.metriportClient = MetriportClient(healthStore: healthStore)
+        
+        // If we've already authorized then start checking background updates on app load
+        if UserDefaults.standard.object(forKey: "HealthKitAuth") != nil {
```
Comment (Goncharo) [inline pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthStore.swift:17]: What happens if the user manually revokes the permissions in settings?

Should we always be checking if permissions are actually still granted just in case?

**Review by Goncharo — CHANGES_REQUESTED** [review]
Generally looks good - great job. Few minor comments in addition to the previous ones

File: examples/ios/WebViewExample/Metriport mobile/Metriport mobile.entitlements [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.developer.healthkit</key>
+	<true/>
+	<key>com.apple.developer.healthkit.access</key>
+	<array/>
+	<key>com.apple.developer.healthkit.background-delivery</key>
+	<true/>
```
[the remark this answers is not in the document]
Comment (Orta21) [reply examples/ios/WebViewExample/Metriport mobile/Metriport mobile.entitlements:10]: We should ill do that as well.

File: examples/ios/HealthKitExample/Metriport mobile.xcodeproj/xcshareddata/xcschemes/Metriport mobile.xcscheme [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            isEnabled = "YES">
+         </EnvironmentVariable>
+         <EnvironmentVariable
+            key = "OSS_API_URL"
+            value = "http://192.168.0.139:8080"
+            isEnabled = "YES">
+         </EnvironmentVariable>
+         <EnvironmentVariable
+            key = "CLIENT_API_KEY"
+            value = "eXV6N0FISEVTMGJ3azhKTVg2bFZBOmY1NjUwYWE0LWRkZWQtNDIwZi05YjM2LWVlMmRiZmNjZTNlYQ"
```
[the remark this answers is not in the document]
Comment (Orta21) [reply examples/ios/HealthKitExample/Metriport mobile.xcodeproj/xcshareddata/xcschemes/Metriport mobile.xcscheme:94]: This is how ENVS get set https://m25lazi.medium.com/environment-variables-in-xcode-a78e07d223ed

I guess when it builds out it creates this file. Ill make sure this doesnt get commited

File: api/app/src/mappings/apple/index.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    !!data.HKQuantityTypeIdentifierRestingHeartRate ||
+    !!data.HKQuantityTypeIdentifierHeartRateVariabilitySDNN ||
+    !!data.HKQuantityTypeIdentifierBodyTemperature ||
+    !!data.HKQuantityTypeIdentifierBloodPressureSystolic ||
+    !!data.HKQuantityTypeIdentifierBloodPressureDiastolic ||
+    !!data.HKQuantityTypeIdentifierRespiratoryRate
+};
+
+
+export const appleItem = z.object({ date: z.string(), value: z.number() })
```
Comment (leite08) [inline api/app/src/mappings/apple/index.ts:98]: We could check the date is in the format specified by Apple and is a valid value.

For reference: https://github.com/metriport/metriport/blob/develop/api/app/src/mappings/garmin/index.ts#L31

File: api/app/src/command/connected-user/get-connected-user.ts [hunk 6]
```diff
@@ -3,6 +3,29 @@ import { ConnectedUser } from "../../models/connected-user";
...
+}: {
+  id: string;
+}): Promise<ConnectedUser | null> => {
+  const connectedUser = await getConnectedUserById({ id });
+  if (!connectedUser)
+    throw new NotFoundError(
+      `Could not find connected user ${id}`
+    );
+  return connectedUser;
+};
```
[the remark this answers is not in the document]
Comment (leite08) [reply api/app/src/command/connected-user/get-connected-user.ts:27]: That, and if the use case warrants querying without a `cxId`, we could modify that method to make `cxId` optional.

File: api/app/src/mappings/apple/activity.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+          }
+        }
+        return;
+      }
+
+      activity.push({
+        metadata: {
+          date: dayjs(appleItem.date).format("YYYY-MM-DD"),
+          source: PROVIDER_APPLE,
+        },
```
Comment (leite08) [inline api/app/src/mappings/apple/activity.ts:32]: Could have a function to return metadata getting a date as parameter to reuse this logic.

File: api/app/src/mappings/apple/activity.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+              basal_metabolic_rate_kcal: appleItem.value
+            }
+          }
+        }
+        return;
+      }
+
+      activity.push({
+        metadata: {
+          date: dayjs(appleItem.date).format("YYYY-MM-DD"),
```
Comment (leite08) [inline api/app/src/mappings/apple/activity.ts:60]: We have a constant for the literal determining the format of Date ISO: `./shared/date.ts/ISO_DATE`

It would be good to replace this and others.

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthKitTypes.swift [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            HKSampleType.quantityType(forIdentifier: .flightsClimbed)!:
+            return .count()
+            
+        case
+            HKSampleType.quantityType(forIdentifier: .activeEnergyBurned)!,
+            HKSampleType.quantityType(forIdentifier: .basalEnergyBurned)!:
+            return .kilocalorie()
+            
+        case HKSampleType.quantityType(forIdentifier: .vo2Max)!:
+            return .literUnit(with: .milli).unitDivided(by: .gramUnit(with: .kilo).unitMultiplied(by: .minute()))
```
[the remark this answers is not in the document]
Comment (Orta21) [reply pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthKitTypes.swift:96]: Its not really a conversion. What this does like for all the other cases is give the unit we would like this to be in. Which in our case is ml/kg/min

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthStore.swift (hunk 2 above)
Comment (Orta21) [reply pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/HealthStore.swift:17]: So the HealthKitAuth key is just to verify that the user has been shown the screen to access the different data types.

The reason we have it is so we dont make requests to healthkit before being that.

What the user can do is remove access to certain data types (or all of them). What will happen then is we just wont receive that data but no errors will be thrown.

File: examples/ios/WebViewExample/Metriport mobile/Metriport mobile.entitlements (hunk 3 above)
[the remark this answers is not in the document]
Comment (Orta21) [reply examples/ios/WebViewExample/Metriport mobile/Metriport mobile.entitlements:10]: Yea forsure once i make the other changes and everything else looks good ill make one last commit with the updated docs

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportApi.swift [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        
+        if let apiUrl = ProcessInfo.processInfo.environment["OSS_API_URL"] {
+            var request = URLRequest(url: URL(string: "\(apiUrl)/webhook/apple")!)
+            request.httpMethod = "POST"
+            request.httpBody = bodyData
+            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            
+            if let clientApiKey = ProcessInfo.processInfo.environment["CLIENT_API_KEY"] {
+                request.addValue(clientApiKey, forHTTPHeaderField: "x-api-key")
+            }
```
[the remark this answers is not in the document]
Comment (Orta21) [reply pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportApi.swift:41]: So definitely valid but i needed to pass it to the initialization of the MetriportHealthStoreManager and then eventually it makes its way to the MetriportApi

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportClient.swift [hunk 11]
```diff
@@ -0,0 +1,351 @@
+import Foundation
+import HealthKit
+import Combine
+import CoreData
+import WebKit
+
+struct ThirtyDaySamples: Codable {
+    var type: String
+    var samples: [Sample]
+}
```
[the remark this answers is not in the document]
Comment (Orta21) [reply pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportClient.swift:10]: I actually dont use this so i removed it

File: pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportWidget.swift [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+}
+
+// This handles the communication between widget and native code
+extension MetriportWidget.Coordinator: WKScriptMessageHandler {
+    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
+        delegate?.receivedJsonValueFromWebView(value: ["data": message.body])
+    }
+}
```
[the remark this answers is not in the document]
Comment (Orta21) [reply pkgs/ios/Metriport-IOS/Sources/Metriport-IOS/MetriportWidget.swift:103]: 🪄

File: api/app/src/mappings/apple/index.ts (hunk 5 above)
Comment (Orta21) [reply api/app/src/mappings/apple/index.ts:98]: So i did this 
  date: z.string().refine((v) => dayjs(v, "YYYY-MM-DDTHH:mm:ss", true).isValid())
  
  No matter what i did though it always returns false but in a repl it works just fine its weird.

File: examples/ios/HealthKitExample/Metriport mobile.xcodeproj/xcshareddata/xcschemes/Metriport mobile.xcscheme (hunk 4 above)
[the remark this answers is not in the document]
Comment (Goncharo) [reply examples/ios/HealthKitExample/Metriport mobile.xcodeproj/xcshareddata/xcschemes/Metriport mobile.xcscheme:94]: just wanted to flag that this is still here, should this be ignored?

File: packages/ios/Metriport-Health/Package.swift [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+    targets: [
+        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
+        // Targets can depend on other targets in this package, and on products in packages this package depends on.
+        .target(
+            name: "Metriport-IOS",
+            dependencies: []),
+        .testTarget(
+            name: "Metriport-IOSTests",
+            dependencies: ["Metriport-IOS"]),
```
Comment (Goncharo) [inline packages/ios/Metriport-Health/Package.swift:27]: This is now called Metriport-Health, but everything else references Metriport-IOS, including some sub-dirs. Is there a proper rename that we should do?

Also, I'd vouch for naming this `metriport-sdk` but not a big deal

File: packages/ios/Metriport-Health/README.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+# Metriport-IOS
+
+A description of this package.
```
Comment (Goncharo) [inline packages/ios/Metriport-Health/README.md:3]: Needs updating

**Review by Goncharo — CHANGES_REQUESTED** [review]
Few minor comments to be addressed -  code LGTM

File: README.md [hunk 15]
```diff
@@ -122,13 +122,23 @@ Our beautiful developer documentation, powered by [mintlify](https://mintlify.co
...
+
+- [Metriport API](/pkgs/npm/packages/api/): contains the Metriport data models, and a convenient API client wrapper.
+- [CommonWell JWT Maker](/pkgs/npm/packages/commonwell-jwt-maker/): CLI to create a JWT for use in [CommonWell](https://www.commonwellalliance.org/) queries.
+- [CommonWell SDK](/pkgs/npm/packages/commonwell-sdk/): SDK to simplify CommonWell API integration.
+
+#### **iOS**
+
+Our iOS packages are available in [`/pkgs/ios`](/pkgs/ios):
+
+- [Metriport iOS](/pkgs/ios/Metriport-IOS): SDK to integrate with the Metriport Connect Widget and Apple Health on iOS.
```
Comment (Goncharo) [inline README.md:141]: these paths will need to be updated for /packages, & the iOS section linked to our other repo

File: docs/more-info/iOS.mdx [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+---
+
+## Initial Setup
+
+**Enable HealthKit:**
+
+First thing you need to do is enable HealthKit capabilities for your app.
+In Xcode, select the project, click on the `Signing & Capabilities` tab and add the HealthKit capability.
+Make sure to select `Background Delivery` under `Capabilities` to be able to
+receive data via webhooks.
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:13]: should we make this one line so the text can automatically break up based on browser width?

File: docs/more-info/iOS.mdx [hunk 17]
```diff
@@ -0,0 +1,133 @@
+---
+title: "iOS"
+description: "Our iOS SDK allows you to receive Apple HealthKit data via webhooks"
+---
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:4]: Should we call this whole section Apple Health instead of iOS? I think people's tendency would be to seek out an Apple Health doc.

File: docs/more-info/iOS.mdx [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+```swift
+<key>NSHealthShareUsageDescription</key>
+<string>Example: Share usage with all people</string>
+```
+
+**Install Package:**
+
+Our packages can be downloaded using SPM (Swift Package Manager).
+To add a package dependency to your Xcode project, select `File > Add Packages` and enter https://github.com/metriport/metriport-ios-sdk.
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:27]: Maybe putting the URL in a snippet would be good for easy copy/paste?

```md
https://github.com/metriport/metriport-ios-sdk
```

File: docs/more-info/iOS.mdx [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+        }
+        .sheet(isPresented: $webviewController.showWebView) {
+            MetriportWidget(url: "\(WIDGET_URL)/?token=\("TOKEN")", healthStore: healthStore)
+        }
+    }
+    .padding()
+}
+```
+
+<Info>Make sure to pass the healthStore to the widget</Info>
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:64]: "to the widget."

File: docs/more-info/iOS.mdx [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+These are the data types that we are requesting access too and receiving from Apple HealthKit:
+
+```swift
+// ACTIVITY
+HKObjectType.quantityType(forIdentifier: .stepCount)!,
+HKObjectType.quantityType(forIdentifier: .activeEnergyBurned)!,
+HKObjectType.quantityType(forIdentifier: .basalEnergyBurned)!,
+HKObjectType.quantityType(forIdentifier: .flightsClimbed)!,
+
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:84]: Perhaps provide an example payload the consumer can expect to receive instead?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: docs/more-info/iOS.mdx [hunk 21]
```diff
@@ -1,16 +1,31 @@
...
-title: "iOS"
-description: "Our iOS SDK allows you to receive Apple HealthKit data via webhooks"
+title: "Apple Health"
+description: "Our iOS SDK allows you to receive Apple HealthKit data via webhooks."
 ---
 
+## Overview
+
+This is a guide to help you get started with iOS, specifically Apple HealthKit. First and foremost Apple HealthKit is a central repository for health and fitness data only available on iPhone and Apple Watch.
+Where Metriport comes in is we provide an SDK that allows you to be able to receive Apple HealthKit data in the background.
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:9]: I'd remove "and Apple Watch" -> you can technically import data from other sources into Apple Health.

Also, should we use the terminology "Apple Health" rather than "Apple HealthKit"? HealthKit is the dev tool, but our cxs are looking to integrate with Health.

File: docs/more-info/iOS.mdx [hunk 22]
```diff
@@ -21,10 +36,14 @@ Also add the following key in your Info.plist:
 <string>Example: Share usage with all people</string>
 ```
 
-**Install Package:**
+### 2. Install Metriport iOS SDK Package
 
 Our packages can be downloaded using SPM (Swift Package Manager).
-To add a package dependency to your Xcode project, select `File > Add Packages` and enter https://github.com/metriport/metriport-ios-sdk.
+To add a package dependency to your Xcode project, select `File > Add Packages` and enter
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:42]: "... and enter our repo URL:"

File: docs/more-info/models.mdx [hunk 23]
```diff
@@ -0,0 +1,409 @@
+---
+title: "Data Models"
+description: "Below is a comprehensive list of our data models that you should expect to see when receiving data."
+---
```
Comment (Goncharo) [inline docs/more-info/models.mdx:4]: good call

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: docs/more-info/iOS.mdx [hunk 24]
```diff
@@ -0,0 +1,98 @@
+---
+title: "Apple Health"
+description: "Our iOS SDK allows you to receive Apple HealthKit data via webhooks."
+---
+
+## Overview
+
+This is a guide to help you get started with iOS, specifically Apple HealthKit. First and foremost Apple HealthKit is a central repository for health and fitness data only available on iPhone and Apple Watch.
+Where Metriport comes in is we provide an SDK that allows you to be able to receive Apple HealthKit data in the background.
+When data gets added to Apple HealthKit by a user it will simply send that data to the Metriport api and from there it gets sent to your webhook url.
```
Comment (Goncharo) [inline docs/more-info/iOS.mdx:10]: URL

**Review by Goncharo — CHANGES_REQUESTED** [review]

**Review by Goncharo — APPROVED** [review]
🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢
